[Zhou Xingye] iP

starrdye opened the pull request:
# Miku project

>連れてって いろんな世界を見たい キミと一緒に  - Hatsune Miku

This is a project done by Xingye for a greenfield Java project. It's named after the famous japnese Idol Hatsune Miku. Given below are instructions on how to use it.

## Setting up in Intellij

Prerequisites: JDK 17, update Intellij to the most recent version.

1. Open Intellij (if you are not in the welcome screen, click `File` > `Close Project` to close the existing project first)
1. Open the project into Intellij as follows:
   1. Click `Open`.
   1. Select the project directory, and click `OK`.
   1. If there are any further prompts, accept the defaults.
1. Configure the project to use **JDK 17** (not other versions) as explained in [here](https://www.jetbrains.com/help/idea/sdk.html#set-up-jdk).<br>
   In the same dialog, set the **Project language level** field to the `SDK default` option.
3. After that, locate the `src/main/java/Duke.java` file, right-click it, and choose `miku.Run Duke.main()` (if the code editor is showing compile errors, try restarting the IDE). If the setup is correct, you should see something like the below as the output:

```
Hello from
 __  __   __   _   _   _    _
|  \/  | |  | | | / / | |  | |
| |\/| | |  | | |/ /  | |  | |
| |  | | |  | | |\ \  | |__| |
|_|  |_| |__| |_| \_\  \____/
Hello! I'm Miku
What song do you want to listen to today?
```

The Run.class is the launcher:
```Java

/**
 * Executes the program.
 */
public class Run {
   ...
}
```

## User Guide
1. Download the .jar file [<u>here</u>](https://github.com/zhou-colla/ip/releases).
2. **Run** the .jar file | **Alternatively** `java -jar Miku.jar.`
3. Let Miku help you with task management! 💚

## Features
- Arrange tasks for you. Add/Delete/Mark/Unmark the following tasks.
   - Deadline
   - Event
   - Todo
- Search the matching tasks from the stored tasks.

## Upcoming Updates
- [x] ~~Create a better userinterface using javaFX.~~
- [ ] Add musicbox for BGM.
- [ ] Add Miku personality to the UI.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

starrdye committed:
Add greet function and farewell function to the chatbot Miku

The class miku now has two functions \n1. Miku.greet(): print greeting message.\n2. Miku.farewell() print farewell message.

starrdye committed:
Remove the Duke chatbot

Remove the Duke.java file

starrdye committed:
Add echo function

starrdye committed:
add list and add functions

starrdye committed:
add task class

add task class with mark, unmark and other functions

starrdye committed:
Added Event, Deadline, Todo

Further extension from task, adding features to add todo/deadline/event

starrdye committed:
task counter after adding a task

starrdye committed:
Added semi auto testing

added a test case to add two todo tasks, delete 1 then exit the program

starrdye committed:
Added RemoveNullException, and disabled printing Miku's face for simplicity

disabled priting mikus face and added an exception to check if delete command is valid

starrdye committed:
Added InvalidCommandException

Added another exception and the logic to check if the user input an valid command

starrdye committed:
formatted the codes, finalised the delete command

finalised miku, delete function and codes

starrdye committed:
Add auto save function to the app

starrdye committed:
Add directory check

starrdye committed:
Merge branch 'branch-Level-7'

starrdye committed:
Add Time Understanding Feature

starrdye committed:
Add .gitignore

starrdye committed:
Merge branch 'branch-Level-8'

starrdye committed:
Add more OOP for storage and ui

starrdye committed:
Add Run.java

starrdye committed:
Remove Miku.java

starrdye committed:
Divide classes into packages

starrdye committed:
Remove Duke.java

starrdye committed:
Merge branch 'add-gradle-support'

starrdye committed:
Update readme file and build.gradle

starrdye committed:
Add RunTest.java

starrdye committed:
Add java docs for exception and command classes

starrdye committed:
Correct the coding standard violations

starrdye committed:
Add FindCommand to enable search

starrdye committed:
Merge branch 'branch-A-JavaDoc'

starrdye committed:
Merge branch 'branch-A-CodingStandard'

starrdye committed:
Remove setdata method

starrdye committed:
Merge branch 'branch-Level-9'

starrdye committed:
Fix some doc formatting issues

starrdye committed:
Correct the coding standard violations

starrdye committed:
Merge branch 'A-CheckStyle'

starrdye committed:
Update README.md

keerthigkaarthik commented on README.md:
````diff
@@ -13,12 +15,44 @@ Prerequisites: JDK 17, update Intellij to the most recent version.
+3. After that, locate the `src/main/java/Duke.java` file, right-click it, and choose `miku.Run Duke.main()` (if the code editor is showing compile errors, try restarting the IDE). If the setup is correct, you should see something like the below as the output:
+
+```
+Hello from
+ __  __   __   _   _   _    _
+|  \/  | |  | | | / / | |  | |
+| |\/| | |  | | |/ /  | |  | |
+| |  | | |  | | |\ \  | |__| |
+|_|  |_| |__| |_| \_\  \____/
+Hello! I'm Miku
+What song do you want to listen to today?
````
Nice! I think this looks really cool.

keerthigkaarthik commented on src/main/java/miku/Run.java:
```diff
@@ -0,0 +1,43 @@
+import miku.parser.CommandMikuParser;
+import miku.utility.Storage;
+import miku.utility.TaskList;
+import miku.utility.UI;
```
I like how you have all of your important functions (commands, etc) separated nicely into their own classes and packages.

keerthigkaarthik commented on src/main/java/miku/command/FindCommand.java:
```diff
@@ -0,0 +1,22 @@
+
+    @Override
+    public void execute(TaskList taskList, UI ui, Storage storage) {
+        taskList.searchList(string);
```
Maybe you meant to save your result into your "results" field?

keerthigkaarthik commented on src/main/java/miku/utility/TaskList.java:
```diff
@@ -0,0 +1,82 @@
+     *
+     * @param string the matching key phrase.
+     */
+    public void searchList(String string) {
+        int counter = 1;
+        for (int i = 0; i < tasks.size(); i++) {
+            if (tasks.get(i).getDesc().contains(string)) {
+                System.out.println(String.valueOf(counter) + ". " + tasks.get(i).stringValue());
+                counter += 1;
+            }
+        }
+    }
```
Perhaps you can return a new task list from this method instead, save it into your "FindCommand" and have your printList method handle this? If you would like to change how the list is printed in the future you would only need to update printList and not both methods.

keerthigkaarthik commented on src/main/java/miku/command/MarkCommand.java:
```diff
@@ -0,0 +1,39 @@
+     * @param storage  The storage function to store the data into a text file
+     */
+    @Override
+    public void execute(TaskList taskList, UI ui, Storage storage) {
+        taskList.mark(index);
+        storage.saveFromTaskList(taskList);
+        ui.printSectionBreak();
+    }
```
You've added the try, catch blocks for invalid indexes in other Command classes, you may want to consider adding it here too.

keerthigkaarthik commented on src/main/java/miku/command/RemoveCommand.java:
```diff
@@ -0,0 +1,47 @@
+     */
+    @Override
+    public void execute(TaskList taskList, UI ui, Storage storage) {
+        try {
+            if (index < 1 || index > taskList.size()) {
+                throw new RemoveNullException("In valid index");
```
Nice!

keerthigkaarthik commented on src/main/java/miku/parser/CommandMikuParser.java:
```diff
@@ -0,0 +1,72 @@
+ * Parses the input command strings
+ */
+public class CommandMikuParser extends MikuParser {
+    private static String regexMark = "mark \\d+";
+    private static String regexUnmark = "unmark \\d+";
+    private static String regexToDo = "todo .*";
+    private static String regexDeadline = "deadline .* /by .*";
+    private static String regexEvent = "event .* /.* /.*";
+    private static String regexRemove = "delete \\d+";
+    private static String regexFind = "find .*";
```
This is really cool! I didn't know you could do this. I split mine up one at a time, e.g (event) input.split.("/from").split("/to") etc

yimjianbing replied:
Woah yeah i agree, I was shocked when I saw this! Really cool!

keerthigkaarthik reviewed (commented):
The code is well written. All the functionality is split up very nicely, and OOP is implemented very thoroughly. I like how much effort you put into the UI with Miku's face and the introduction line.

yimjianbing commented on src/main/java/miku/command/UnmarkCommand.java:
```diff
@@ -0,0 +1,35 @@
+ * Unmarks a specified task.
+ */
+public class UnmarkCommand extends Command {
+    int index;
```
I think you forgot to add a private modifier to this field, but other than that the rest looks great!

yimjianbing replied:
I think you forgot to add a private modifier to this field, but other than that the rest looks great!

yimjianbing commented on src/main/java/miku/exception/InvalidCommandException.java:
```diff
@@ -0,0 +1,24 @@
+    }
+
+    /**
+     * prints the error with a description.
```
SUPER small nit but the word "print" in the javadocs should be capitalised

yimjianbing commented on src/main/java/miku/task/Event.java:
```diff
@@ -0,0 +1,74 @@
+/**
+ * Represents an event object.
+ */
+public class Event extends Task {
+    private String dateTime1 = "";
+    private String dateTime2 = "";
+    private LocalDateTime startDateTime;
+    private LocalDateTime endDateTime;
```
Great job on integrating the LocalDateTime java package into your Event class as well! I didn't do it so props to you!

yimjianbing commented on src/main/java/miku/utility/UI.java:
```diff
@@ -0,0 +1,77 @@
+public class UI {
+    private String sectionBreak = "\n_______________________________________________";
+    private Scanner scanner;
+    private String mikuFace = "⠀⠀⠀⠀⠀⠀⢀⡤⣢⠟⢁⣴⣾⡿⠋⢉⠱⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⠉⠑⠒⠦⢄⣀⣴⠟⢡⣠⣼⣿⡿⢳⣄⡀⠀⠀\n"
+            + "⠀⠀⠀⠀⠀⢀⣾⡿⠃⣠⣿⣿⠿⠂⠀⠉⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢲⡿⠋⢰⣾⣿⣿⡟⠀⠀⠈⠙⢆⠀\n"
+            + "⠀⠀⠀⠀⠀⡜⠻⣷⣾⣿⠟⠁⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢰⣿⣁⣰⢸⣿⢻⠟⢀⠀⠀⠀⠀⠀⠁\n"
+            + "⠀⠀⠀⠀⠰⠀⠀⢙⡿⠁⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢻⣿⣿⣿⣿⣯⡀⠀⢃⠀⠀⠀⠀⠀\n"
+            + "⠀⠀⠀⠀⠀⠀⢠⠎⠀⠀⠀⠀⠀⠀⣼⠀⢀⠀⠀⠀⠀⠀⢠⣷⡀⠀⠀⠀⠀⡀⠄⠀⠀⠀⠀⢻⣿⣿⣿⣧⠑⠀⣢⡄⠀⠀⠀⠀\n"
+            + "⠀⠀⠀⠀⠀⡰⠃⢀⠄⠀⠀⠀⠀⣼⡿⡆⢸⠀⠀⠀⠀⠀⠈⣿⢷⡄⠀⠀⠀⠱⡀⠰⡀⠀⠀⠈⢿⣿⣿⣿⣧⠀⢸⣧⠀⠀⠀⠀\n"
+            + "⠀⠀⠀⠀⠀⠡⢢⠋⠀⠀⠀⠀⣼⡟⠀⣇⢸⡆⠀⠀⠀⡄⠀⢿⠀⢳⡄⠀⠀⠀⢳⠀⢳⠀⠀⠀⠈⣿⣿⣿⣿⣷⣘⡟⠀⠀⠀⠀\n"
+            + "⠀⠀⠀⠀⠀⢀⠇⠀⠀⠀⠀⣸⡿⢤⠤⠸⡸⣷⠀⠀⠀⢱⠀⣾⡤⠤⢿⡤⢀⡀⠀⢧⠘⡆⠀⠀⠀⢸⡟⠻⢿⠟⣿⣷⡄⠀⠀⠀\n"
+            + "⠀⠀⠀⠀⠀⡞⠀⠀⠀⠀⢰⡿⢠⠇⠀⠀⢳⣿⢇⠀⠀⠈⡇⣿⡇⠀⠀⠻⣄⠀⠀⠘⡆⡇⠀⠀⠀⠀⣇⢀⡏⠀⣿⡿⣄⠀⠀⠀\n"
+            + "⠀⠀⠀⠀⢰⠁⠀⠀⠀⠀⣿⠁⣄⣀⣀⡀⠈⢿⡜⡄⠀⠀⢹⣿⡇⠐⢄⣀⠘⢧⡀⠀⠹⣿⠀⠀⠀⠀⢸⣿⣷⣶⣿⡇⢹⡇⠀⠀\n"
+            + "⠀⠀⠀⠀⠾⠀⠀⠀⠀⢸⣧⣾⠟⢉⣽⣿⣦⠈⢷⡘⣆⠀⠸⡟⣷⣶⠟⠛⢻⣷⣦⣀⠀⢻⠀⠀⠀⠀⢸⣏⣩⣼⣿⡇⠈⣷⠀⠀\n"
+            + "⠀⠀⠀⠃⠀⠀⠀⠀⠀⣿⡿⠁⠀⣠⣾⣿⣿⠀⠈⢿⠺⡆⠀⣧⢸⠀⠀⢀⣹⣿⣿⣿⣷⣼⣤⠀⠀⠀⢸⣿⣿⣿⣿⠀⠀⣿⠀⠀\n"
+            + "⠀⠀⣠⠄⣀⠀⠀⠀⢠⣿⡇⠀⠀⢻⢻⣟⢿⠀⠀⠈⠣⠈⠓⠾⠀⠀⠀⣿⣿⢿⣿⣿⠘⡇⡞⠀⠀⢠⣾⣿⣿⣿⡏⠀⠀⢹⠀⠀\n"
+            + "⠀⠀⠛⠀⣿⠀⠀⠀⢸⣿⣿⡀⠀⠈⠃⠐⠋⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠸⣄⣐⣠⠏⢠⣿⠁⠀⠀⢸⣿⣿⣿⣿⠀⠀⠀⢸⠀⠀\n"
+            + "⠀⠀⠀⠀⢹⡆⠰⡀⢸⡟⠩⣷⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣼⠃⠀⠀⠀⢸⣿⣿⣿⠟⠀⠀⠀⠘⠀⠀\n"
+            + "⠀⠀⠀⠀⢎⣿⡀⢱⢞⣁⣀⡿⡄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣠⠞⡏⡼⠀⠀⠀⣾⣿⠋⠁⠀⠀⠀⠀⠀⠀⠀\n"
+            + "⠀⠀⠀⠀⠈⠿⠻⡇⠀⠀⠒⠢⢵⡄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢰⣽⠁⠀⠀⢠⡿⢹⠀⠀⠀⠀⠀⠀⠀⠀⠀\n"
+            + "⡀⠀⠀⠀⠀⠀⠀⡟⣦⡀⠀⠀⠀⠈⠓⢄⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣀⣴⢿⡇⠀⠀⡄⣸⣇⣼⣀⣀⣀⠀⠀⠀⠀⠀⠀\n"
+            + "⡀⠀⠀⠀⠀⠀⢰⠇⣿⢸⣦⡀⠀⠀⠀⠀⠈⠲⣄⡀⠀⠀⠀⠀⠀⣀⡤⠒⢉⡴⠃⣸⠀⠀⢶⣿⣿⡿⠃⡤⠊⠁⠉⠑⢄⠀⠀⠀\n"
+            + "⡇⠀⠀⠀⠀⠀⢸⠀⣿⣾⣿⢿⠲⣄⠀⠀⠀⠀⠘⠟⣦⣤⣴⡒⠉⢀⡠⠖⠉⠀⣠⠃⠀⣠⣿⣿⡿⠁⠊⠀⠀⠀⠀⠀⠀⠀⠀⠀\n"
+            + "⠀⠀⠀⠀⠀⠀⢸⠀⣿⠛⢿⠈⢢⠏⠀⠀⠀⠀⠀⣰⣏⣀⣿⠗⠊⠁⠀⠀⣠⣾⠃⢀⡴⠿⠛⣿⠁⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀\n"
+            + "⠀⠀⠀⠀⠀⠀⢸⢀⠇⠀⠈⢠⠃⠀⠀⠀⠀⠀⢰⠟⠁⠀⢹⢇⠀⣀⠴⠊⡱⠥⠔⠋⠀⠀⢠⠏⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀\n"
+            + "⠀⠀⠀⠀⠀⠀⢸⡟⠀⢀⡴⠁⠀⠀⠀⠀⠀⢠⡟⠀⠀⣰⢿⡘⣾⡅⠀⠀⠀⠀⢀⠄⠀⢠⠏⢀⣄⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀\n"
+            + "⠀⠀⠀⠀⠀⠀⢸⠀⣰⣿⠀⠀⠀⠀⠀⠀⢠⣿⠃⢀⡾⡇⠘⠻⡿⢷⡀⠀⠀⠒⠁⠀⢠⠏⢀⠏⣸⠃⢻⠏⠀⠀⠀⠀⠀⠀⠀⠀\n"
+            + "⠀⠀⠀⠀⠀⠀⣧⣾⣹⣿⠀⠀⠀⠀⠀⢠⠏⢉⠀⡞⣰⡇⠀⣴⣥⠞⢷⠀⠀⠀⠀⣠⠎⠀⠸⣶⠋⣠⡟⠀⠀⠀⠀⠀⠀⠀⠀⠀";
```
really cool!

yimjianbing commented on src/main/java/miku/task/Task.java:
```diff
@@ -0,0 +1,76 @@
+     *
+     * @param isDone
+     */
+    public void isDoneInit(boolean isDone) {
```
Maybe a name like initIsDone wuld be more appropriate as a name to following the coding standard?

yimjianbing reviewed (commented):
Overall I think the code looks great! Everything is nicely organised and i can easily navigate through the codespace and understand what every class and function does so good job!

Some nits here and there, but everything else is amazing!

starrdye committed:
Add GUI for the chatbot

starrdye committed:
Merge remote-tracking branch 'origin/master'

starrdye committed:
Correct minor coding standard violations

starrdye committed:
Merge branch 'branch-Level-10'

starrdye committed:
Delete unused methods

starrdye committed:
Add assertions to ensure command and response integrity

The code does not currently check for null commands returned by the parser or validate that the UI generates a response after executing a command.

Failing to validate these conditions could result in unexpected null pointers or incomplete executions, leading to crashes or incorrect behavior. By ensuring that commands and responses are valid, we can catch issues early in development and improve the robustness of the program.

Add assertions to check for null commands returned by the parser or validate that the UI generates a response after executing a command:

Assertions are used because they are lightweight checks during development that can help catch issues early.

starrdye committed:
Merge pull request #1 from zhou-colla/branch-A-Assertions

Add assertions to ensure command and response integrity

starrdye committed:
Merge branch 'master' into branch-A-CodeQuality

starrdye committed:
Merge pull request #2 from zhou-colla/branch-A-CodeQuality

Delete unused methods

starrdye committed:
Add Priority field to Task class and create Priority enum

The Task class, along with its subclasses Deadline, Todo, and Event, currently do not support any way to assign or track task priorities.

Users are unable to prioritize tasks, which limits their ability to manage tasks based on importance or urgency. Without a priority system, tasks are treated equally, reducing flexibility in task organization.

Add a priority field to the Task class, which is inherited by Deadline, Todo, and Event. Create an enum Priority to represent different levels of priority (e.g., HIGH, MEDIUM, LOW). Update the task creation and display logic to incorporate this new field.

Adding the priority field to the Task class allows all subclasses to inherit the functionality without repeating code. The Priority enum ensures that priority levels are standardized and easy to manage. This also supports future changes and expansions, as adding more priority levels or modifying behavior becomes simpler.

starrdye committed:
Merge branch 'add-priority-to-task'

The Task class, along with its subclasses Deadline, Todo, and Event, currently do not support any way to assign or track task priorities.

Users are unable to prioritize tasks, which limits their ability to manage tasks based on importance or urgency. Without a priority system, tasks are treated equally, reducing flexibility in task organization.

Add a priority field to the Task class, which is inherited by Deadline, Todo, and Event. Create an enum Priority to represent different levels of priority (e.g., HIGH, MEDIUM, LOW). Update the task creation and display logic to incorporate this new field.

Adding the priority field to the Task class allows all subclasses to inherit the functionality without repeating code. The Priority enum ensures that priority levels are standardized and easy to manage. This also supports future changes and expansions, as adding more priority levels or modifying behavior becomes simpler.

starrdye committed:
Create gradle.yml

starrdye committed:
Update gradle.yml

starrdye committed:
Update Runtest

starrdye committed:
Fix inconsistent Javadoc comments in command classes

The Javadoc comments in AddCommand.java, Command.java, and ExitCommand.java currently do not follow the required format. They fail to start with a verb in the present tense and do not provide concise summaries, which affects clarity and adherence to the coding standard.

This needs to change because the coding standard specifies that the first sentence of Javadoc comments must be a short summary starting with a verb like "Returns," "Initializes," etc., to ensure proper documentation in the method summary table.

Update the Javadoc comments to use present-tense verbs and provide more descriptive first sentences. Ensure that each comment is concise and adheres to the standard.

This approach ensures that the documentation is both clear and correctly formatted for automatic inclusion in the method summary table. It also improves the maintainability and readability of the code.

starrdye committed:
Correct coding violations

starrdye committed:
Update README.md

starrdye committed:
Update README.md

starrdye committed:
Update README.md

starrdye committed:
Fix display issues

starrdye committed:
Merge branch 'master' of github.com:zhou-colla/ip

starrdye committed:
Fix not scrolling to the bottom issue

starrdye committed:
Add the ui picture

starrdye committed:
Update README.md

starrdye committed:
Fix not scrolling to bottom issue for Linux users
